### DISCUSSION TITLE
大佬们，敝人搭建的vless+nginx+cdn报了400，但配置之类都是正常的

### DISCUSSION THREAD

**Discussion by Xingeqwd** [XTLS/Xray-core#2547] [Q&A] [answered]
架构大概是vless+ws+tls+web+cdn，算是能过CDN的方案中最安全的了吧（XTLS和reality都不能过）
——虽然WS下的VLESS也被破解了吧，VMESS有双层加密，特征码比较明显

**（后面听到某位大佬的说法，才知道原来还有更新的GRPC）**

也只能说是相对而言，毕竟torjan，ss，ssr之类也都不行了。。。
只好选用一下VLESS

顺便提一下环境方面问题吧（应该是没影响）
系统是almalinux，v2ray服务端客户端版本都很新，nginx版本也很新
空间充足，端口没有冲突，网络都是通的。。。

————————
### 以下是问题内容

8.6 VMESS彻底被封之后，8.13成功切换到VLESS，接通了，也有正常访问记录，后面又挂了
9月初换了一个域名，再搭一次，直接就不通了
（只有修改v2ray和nginx配置文件中的域名内容，按理来说不应该啊）

做过了各种各样的排障除错措施，各方面资源资料也都查了很多，Google起码翻了得有几十个网页
就是没法把他解决掉（这么多年的运维真是白做了）因此来找各位大佬求教一下

**这个应该是直接可以排除GFW和CDN问题**，因为nginx日志里面有接收进来了，报错400
【nginx日志如图】
**只有报告400，没有任何有效信息**
![image](https://github.com/XTLS/Xray-core/assets/46107662/b6b41542-8aeb-4b05-976b-4ff310b9938e)

**v2ray那边没有任何的反应**，
lsof -i 和netstat -ano 之类的端口指令都可以看得到v2ray的端口监听
8.13之后只有error.log中有启动的记录，access.log在8.13之后根本没有变更
【v2ray日志文件如图】
![image](https://github.com/XTLS/Xray-core/assets/46107662/5aedac43-8003-4199-a091-6ef2ff31f744)
————————
【v2ray日志内容如图】，上一层窗口为error，下一层窗口为access（还能看得到8.13最后一次接通时GFW导致的错误内容）：
![image](https://github.com/XTLS/Xray-core/assets/46107662/dafe79cd-419a-43ef-872e-71b1a8732f6b)

——————————————
### 以下是配置文件部分
**nginx配置文件**
跟官方文档对了很多遍，几乎是一模一样了
（v2ray的path是一个正常站点下加的location）
`nginx -t`和`nginx -s reload`时候一切正常，语法没问题

```
server {
    listen 443 ssl http2;
    listen 80 ;
    server_name 域名略过了;
    charset utf-8;

    ssl_protocols TLSv1.1 TLSv1.2 TLSv1.3;
    ssl_prefer_server_ciphers on;
          #此处证书文件内容省略了
    ssl_certificate //fullchain.pem;
    ssl_certificate_key //privkey.pem;

    location / {
      #主站点的内容，此处省略了
    }

    location [v2ray站点的path]{
      proxy_redirect off;
      proxy_pass http://127.0.0.1:12345;
      proxy_http_version 1.1;
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection "upgrade";
      proxy_set_header Host $http_host;
      # Show real IP in v2ray access.log
      proxy_set_header X-Real-IP $remote_addr;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      #proxy_intercept_errors on;
      #error_page 400 = [400错误的伪装跳转站点网址];
      #因为加了这个错误页面跳转，错误内容会变成302，此处简化掉了，当作错误是400
    }
}
```

——————————————————
**下为v2ray配置文件内容**
（test和start时候没有报错，语法上是没问题的，这里做了一些简化）
```
{
  "log": {
    "loglevel": "debug",
    "access": "/var/log/v2ray/access.log",
    "error": "/var/log/v2ray/error.log"
  },
  "inbounds": [
    {
      "port": 12345,
      "protocol": "vless",
      "settings": {
        "clients": [
          {
            "email": "zhangsan",
            "id": "xxxxxxxxxxx",
            "level": 1,
            "alterId": 56
          }
        ]
      },
      "streamSettings": {
        "network": "ws",
        "security": "tls",
        "wsSettings": {
          "path": "[v2ray的path，和nginx中的一模一样]",
          "headers": {
            "Host": "[VPN所用域名]"
          }
        },
        "tlsSettings": {    #当然也有试过删掉tlssetting栏位，依然无效，还是400
          "serverName": "[这边写的是VPN域名的主域名]",
          "certificates": [{
            "certificateFile": "证书文件位置/fullchain.pem",
            "keyFile": "证书文件位置/privkey.pem"
          }]
        }
      },
      "listen": "127.0.0.1"#安全方面考量，设置为只监听本地，这应该也没影响吧
    }
  ],
  "outbounds": [
    {
      "protocol": "freedom",
      "settings": {}
    },
    {
      "protocol": "blackhole",
      "settings": {},
      "tag": "blocked"
    }
  ],
  "routing": {
    "rules": [
      {
        "type": "field",
        "ip": [
          "geoip:private"
        ],
        "outboundTag": "blocked"
      }
    ]
  }
}
```

真是非常的匪夷所思啊。。。。配置应该没问题啊，日志又不反应
推测会不会是nginx和v2ray中间的什么地方。。。
听说是v2ray设置里面的streamSettings不能带tls？之前也删过了，没什么效果
如果有大佬帮忙解决问题，我给他一些红包作为酬谢吧

**Comment by chika0801** [comment, marked as answer]
https://github.com/XTLS/Xray-core/issues/1750#issuecomment-1459469821
https://github.com/XTLS/Xray-core/issues/1750#issuecomment-1459340564

不要用 WS+TLS 了，再加上用CDN额外附加减速效果。你爱惜VPS的IP和端口干嘛了。

非要用CDN，用gRPC。都来这里问了，用最新的不好嘛？

**Reply by Xingeqwd** [reply]
我丢，孤陋寡闻了，原来新协议还有能过CDN的啊，我还以为能过CDN的只有这个了。。。
感谢大佬提点，我去研究看看

**Comment by Xingeqwd** [comment]
我丢！感谢某位大佬的提醒。。。大意了，没想到竟然暴漏了一些敏感信息，千算万算。。。
所幸是看到了他回复的提醒。。及时改过来了。不过可惜，好像他的帖子被谁删掉了
（是看到了邮箱的提醒记录。。。可惜待得进来一看，他的回复没了啊）

**Reply by nursery01** [reply]
我自己刪的，我只不過感到不佳

**Reply by Xingeqwd** [reply]
原来如此，哈哈，感谢大佬，不小心搞大意了

**Comment by Xingeqwd** [comment]
卧槽！大佬威武啊，转投grpc直接通了！美滋滋！诸位后来者。。如果跟我一样，转投GRPC罢！NBNBNB!!!!!!!!!!

**Comment by RPRX** [comment]
> 虽然VLESS也是漏洞满满

我们现在正在解决 VLESS “可以使用” 的 BUG，欢迎体验 main 分支

**Reply by Xingeqwd** [reply]
不好意思大佬，无意冒犯，敝人说的有失偏颇了，其实VMESS和VLESS系列也挺好用的，只是TLS in TLS和WS的协议被破解了而已，GRPC还算比较坚挺的（这就去把原文改一下）
（其实我没看出来大佬是不是反串，只要知道大佬是maintainer就好了，请容鄙人说一句，大佬搞开发维护辛苦了！）
